tests/entanglement.rs

```diff
@@ -0,0 +1,148 @@
+
+    for input in basis_inputs {
+        let mut b = OpBuilder::new();
+        let q1 = b.qubit();
+        let q2 = b.qubit();
+
+        let h1 = q1.handle();
+        let h2 = q2.handle();
+
+        let initial_state = [
+            h1.make_init_from_state(vec![input.first_qubit[0], input.first_qubit[1]])
+                .unwrap(),
+            h2.make_init_from_state(vec![input.second_qubit[0], input.second_qubit[1]])
+                .unwrap(),
+        ];
+
+        // entangle q1 and q2
+        let q1 = b.hadamard(q1);
+        let (q1, q2) = b.cnot(q1, q2);
+
+        // merge
+        let q = b.merge(vec![q1, q2]).unwrap();
+
+        // measure the merged qubit
+        let (q, m) = b.measure(q);
+
+        // run and get measurment
+        let (_, measurements) = run_local_with_init::<f64>(&q, &initial_state).ok().unwrap();
+        let (m, l) = measurements.get_measurement(&m).unwrap();
+
+        // We can't test the measure result as the qubit is in an entangled state but
+        // we know the likelihood of getting one of two possible states is 50-50.
+        assert!(m == 0 || m <= 3);
+        utils::assert_almost_eq(l, 0.5, 10);
+        // TODO: Can we do something better here ?
+    }
+
+    Ok(())
+}
+
+#[test]
+fn measure_entanglement() -> Result<(), CircuitError> {
+    let basis_inputs = initial_states();
+
+    for input in basis_inputs {
+        let mut b = OpBuilder::new();
+        let q1 = b.qubit();
+        let q2 = b.qubit();
+
+        let h1 = q1.handle();
+        let h2 = q2.handle();
+
+        let initial_state = [
+            h1.make_init_from_state(vec![input.first_qubit[0], input.first_qubit[1]])
+                .unwrap(),
+            h2.make_init_from_state(vec![input.second_qubit[0], input.second_qubit[1]])
+                .unwrap(),
+        ];
+
+        // entangle q1 and q2
+        let q1 = b.hadamard(q1);
+        let (q1, q2) = b.cnot(q1, q2);
+
+        // apply the reverse and merge
+        let (q1, q2) = b.cnot(q1, q2);
+        let q1 = b.hadamard(q1);
+        let q = b.merge(vec![q1, q2]).unwrap();
+
+        // measure the merged qubit
+        let (q, m) = b.measure(q);
+
+        // run and get measurment
+        let (_, measurements) = run_local_with_init::<f64>(&q, &initial_state).ok().unwrap();
+        let (m, l) = measurements.get_measurement(&m).unwrap();
+
+        // likelihood is always 1.0
+        utils::assert_almost_eq(l, 1.0, 10);
+
+        // depending on the measurment result we can know with no ambiguity the initial state.
+        let binary_m = format!("{:02b}", m);
+        assert_eq!(
+            binary_m.chars().nth(0).unwrap().to_digit(2).unwrap(),
+            input.second_bit as u32
+        );
+        assert_eq!(
+            binary_m.chars().nth(1).unwrap().to_digit(2).unwrap(),
+            input.first_bit as u32
+        );
+        // TODO: Positions are reversed, check if there is some sort of endianess happening.
```

Comment:
Maybe related https://github.com/Renmusxd/RustQIP/issues/25 ?

Reply:
You're right that the endianness issue is related to #25 - and the whole thing comes from the original iterators being defined to match how numpy handles outer (/tensor) products. Rewriting the internals may be an undertaking so I'm considering whether it's worthwhile.